lib/cocoapods-core/podfile/dsl.rb

```diff
@@ -334,6 +334,52 @@ def target(name, options = nil)
         self.current_target_definition = parent
       end
 
+      # Adds a script phase to be integrated with this target. A script phase can be used to execute an arbitrary
+      # script that can use all Xcode environment variables during execution. A target may include multiple script
+      # phases which they will be added in the order they were declared. Deleting a script phase will effectively remove
+      # it from the target if it has been added previously.
+      #
+      # @example
+      #   script_phase :name => 'HelloWorldScript', :script => 'echo "Hello World"'
+      #
+      # @example
+      #   script_phase :name => 'HelloWorldScript', :script => 'puts "Hello World"', :shell_path => '/usr/bin/ruby'
+      #
+      # @param    [Hash] options
+      #           the options for this script phase.
+      #
+      # @option   options [String] :name
+      #           the name of the script phase. This option is required.
+      #
+      # @option   options [String] :script
+      #           the body of the script to execute. This option is required.
+      #
+      # @option   options [String] :shell_path
+      #           the shell path to use for this script phase, for example `/usr/bin/ruby` to use Ruby for this phase.
+      #
+      # @option   options [Array<String>] :input_paths
+      #           the input paths to use for script. This is used by Xcode to determine whether to re-execute this
+      #           script phase if the input paths have changed or not.
+      #
+      # @option   options [Array<String>] :output_paths
+      #           the output paths to use for script. This is used by Xcode to avoid re-executing this script phase if
+      #           none of the output paths have changed.
+      #
+      # @option   options [Boolean] :show_env_vars_in_log
+      #           whether this script phase should output the environment variables during execution.
+      #
+      # @option   options [Symbol] :execution_order
```

Comment:
@segiddins @DanToml I am curious on what you guys think this API should be...

1) Should we even handle this case for starters?
2) If yes, how have you imagined it to work in terms of before or after compile sources, what about other phases like linking or headers? Could the API become `execute_before => 'Compile Sources'` instead?
3) Another option could be to just let users specify and index `index => 1` and we just insert that where the user asked...no questions and no special logic.

Reply:
Hmmm, I tried this today:
- Added some `script_phase`s
- run `pod install`
- move phases to correct position
- run `pod install` again

The custom positions are lost. CocoaPods should either keep the positions, or provide a way of specifying the positions, I think?

Reply:
You mean the custom shell script positions are lost? If confirmed, can you please provide a sample project and file an issue with the project attached?

Reply:
Jep, that's what I mean. I'll try to create a small sample project.

Reply:
@djbe ah, I think I see what you mean. If you mean you manually readjust a script phase after you've added but then it *still* gets reordered after another `pod install` then you will need a `post_install` hook to arrange it. This is how it landed on master instead.

We can definitely do better but I think I now understand what you mean. Please file this in https://github.com/CocoaPods/CocoaPods so we do not comment on this merged PR further and I can mark it as an enhancement.